docs/hugo/content/design/ADR-2024-07-Dynamic-Export-To-Secret-Or-ConfigMap.md

```diff
@@ -0,0 +1,428 @@
+  including defining our own if we want.
+* Security: Users have little/no flow-control constructs, and CEL is hardened for use in multitenant environments such 
+  as Kubernetes APIServer.
+* Power: We can parse the CEL expression and examine the AST ourselves to extract useful information such as 
+  which secrets they're attempting to use, which enables us to deliver on the "no secret APIs called if no secrets used"
+  goal.
+
+**Cons**
+
+* May not be as familiar to users.
+* No string interpolation, which makes the templates not quite as clean as they otherwise could be.
+  * Mitigation: We could write our own, although then that function will _definitely_ not be familiar to users. 
+
+#### Option 3: Write our own simple text-replacement engine
+
+Supporting a format similar to `text/template`, but without any of the other looping/flow-control features.
+`DefaultEndpointsProtocol=https;AccountName={{.Spec.AzureName}};AccountKey={{.Spec.OperatorSpec.Secrets.Key1}};EndpointSuffix=core.windows.net`
+
+**Pros**
+
+* Theoretically simple
+
+**Cons**
+
+* More work than using something off the shelf. Including things like getting errors right, etc.
+* Not as easily extensible as something like CEL if we wanted to add more capabilities.
+* Trusting our security/hardening is scarier than trusting a "real" product.
+
+### Location
+
+#### Option 1: Part of operatorSpec.secrets and operatorSpec.configMap
+
+A new `formattedValues` option in `operatorSpec.secrets` and `operatorSpec.configMaps`
+
+**Secrets**
+```yaml
+  operatorSpec:
+    secrets:
+      hostName:
+        name: redis-secret
+        key: hostName
+      sslPort:
+        name: redis-secret
+        key: port
+      formattedValues:
+        - name: redis-secret # Name of the destination secret
+          key: hostPort # Name of the key in the secret
+          value: "'%s:%s'.format([secret.hostName, secret.sslPort])" # Value (format) of the secret
+```
+
+**ConfigMaps**
+```yaml
+  operatorSpec:
+    configMaps:
+      formattedValues:
+        - name: account-data # Name of the destination configMap
+          key: id # Name of the key in the configMap
+          value: "resource.status.id" # Value (format) of the configMap entry
+```
+
+**Pros**
+
+* Fits into our existing model relatively well.
+
+**Cons**
+
+* Need to deal with possibility of property name collisions with `formattedValues`,
+  though practically it seems unlikely to be common.
+
+#### Option 2: Two new operatorSpec properties: operatorSpec.formattedSecrets and operatorSpec.formattedConfigMaps
+
+Other than the location, this is very similar to the
+[first proposal](#part-of-operatorspecsecrets-and-operatorspecconfigmap).
+
+**Secrets**
+```yaml
+  operatorSpec:
+    secrets:
+      hostName:
+        name: redis-secret
+        key: hostName
+      sslPort:
+        name: redis-secret
+        key: port
+    formattedSecrets:
+      values:
```

Comment:
My reasoning for `values` was that it gives us more extension flexibility in the future if this is an object rather than a map, but maybe that's not a big deal because we have operatorSpec already.

Reply:
I've changed this to remove the intermediate `values`